sympy/matrices/matrices.py

```diff
@@ -2402,18 +2400,24 @@ def dot(self, b):
                     type(b))
 
         mat = self
-        if mat.cols == b.rows:
-            if b.cols != 1:
-                mat = mat.T
-                b = b.T
-            prod = flatten((mat * b).tolist())
-            if len(prod) == 1:
-                return prod[0]
-            return prod
-        if mat.cols == b.cols:
-            return mat.dot(b.T)
-        elif mat.rows == b.rows:
-            return mat.T.dot(b)
+        m = len(mat)
+        n = len(b)
+        if m == n:
```

Comment:
This can be reworked to have fewer nested if statements.  Something like,

```
if 1 not in mat.shape or 1 not in b.shape:
    raise TypeError...
if len(mat) != len(b):
   raise TypeError...

<rest of code>
```

Reply:
yup.. That seems better!!